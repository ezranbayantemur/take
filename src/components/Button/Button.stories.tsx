import React from 'react';
import {Alert, View} from 'react-native';
import {Input} from '@components';
import Button from './Button';
import type {Meta, StoryObj} from '@storybook/react-native';
import type {ButtonProps} from './Button.types';

const ButtonMeta: Meta<ButtonProps> = {
  argTypes: {
    onPress: {
      action: 'on press',
    },
  },
  args: {
    text: 'Button',
    loading: false,
    disabled: false,
    onPress: () => Alert.alert('Button', 'Pressed action'),
  },
  title: 'Button',
  component: Button,
};

export default ButtonMeta;

export const Basic: StoryObj<ButtonProps> = {};

export const PressAction: StoryObj<ButtonProps> = {
  decorators: [
    (_, options) => {
      const [text, setText] = React.useState('');

      function handlePress() {
        Alert.alert('Button', text);
      }

      return (
        <View>
          <Input onChangeText={setText} />
          <Button {...options} text="Alert Text" onPress={handlePress} />
        </View>
      );
    },
  ],
};
