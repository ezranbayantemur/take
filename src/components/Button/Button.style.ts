import {StyleSheet} from 'react-native';
import {color, padding, font, typography} from '@style';

const base = StyleSheet.create({
  container: {
    backgroundColor: color.MONZA,
    alignItems: 'center',
    padding: padding.MEDIUM,
    borderRadius: 10,
    margin: padding.SMALL,
  },
  text: {
    color: color.ALABASTER,
    fontFamily: font.BOLD,
    fontSize: typography.NORMAL,
  },
});

export default {
  default: base,
  disabled: StyleSheet.create({
    container: {
      ...base.container,
      backgroundColor: color.MERCURY,
    },
    text: {
      ...base.text,
      color: color.SMOKE,
    },
  }),
};
