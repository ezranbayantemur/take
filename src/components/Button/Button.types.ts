import type {TouchableOpacityProps} from 'react-native';

export interface ButtonProps extends TouchableOpacityProps {
  text: string;
  loading?: boolean;
}
