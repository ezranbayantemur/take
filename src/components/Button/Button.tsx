import React from 'react';
import {Text, ActivityIndicator, TouchableOpacity} from 'react-native';
import {ButtonProps} from './Button.types';
import styles from './Button.style';

const Button = ({
  text,
  type = 'default',
  loading,
  testID,
  onPress,
  disabled,
  ...rest
}: ButtonProps) => {
  const styleKey = disabled ? 'disabled' : type;
  const handlePress = () => {
    if (loading || disabled || !onPress) {
      return;
    }

    onPress();
  };

  return (
    <TouchableOpacity
      testID={`${testID}_button_touchable`}
      style={styles[styleKey].container}
      disabled={loading || disabled}
      onPress={handlePress}
      {...rest}>
      {loading ? (
        <ActivityIndicator
          testID={`${testID}_button_loading_indicator`}
          color="white"
        />
      ) : (
        <Text testID={`${testID}_button_text`} style={styles[styleKey].text}>
          {text}
        </Text>
      )}
    </TouchableOpacity>
  );
};

export default React.memo(Button);
