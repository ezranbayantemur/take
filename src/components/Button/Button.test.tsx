import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import Button from './Button';

let wrapper: ReturnType<typeof render>;

describe('Button unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <Button testID="basetest" text="test title" onPress={jest.fn} />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render title', () => {
    const title = wrapper.queryByText('test title');
    expect(title).not.toBeNull();
  });

  it('should render loading indicator if given', () => {
    const _wrapper = render(
      <Button testID="basetest" text="" loading onPress={jest.fn} />,
    );

    const loading = _wrapper.queryByTestId('basetest_button_loading_indicator');
    expect(loading).not.toBeNull();
  });

  it('should trigger onPress correctly', () => {
    const mockOnPress = jest.fn();
    const _wrapper = render(
      <Button testID="basetest" text="" onPress={mockOnPress} />,
    );

    const touchable = _wrapper.getByTestId('basetest_button_touchable');
    fireEvent(touchable, 'onPress');

    expect(mockOnPress).toBeCalled();
  });

  it('should not trigger onPress on loading', () => {
    const mockOnPress = jest.fn();
    const _wrapper = render(
      <Button testID="basetest" text="" loading onPress={mockOnPress} />,
    );

    const touchable = _wrapper.getByTestId('basetest_button_touchable');
    fireEvent(touchable, 'onPress');

    expect(mockOnPress).not.toBeCalled();
  });

  it('should not trigger onPress on disabled', () => {
    const mockOnPress = jest.fn();
    const _wrapper = render(
      <Button testID="basetest" text="" disabled onPress={mockOnPress} />,
    );

    const touchable = _wrapper.getByTestId('basetest_button_touchable');
    fireEvent(touchable, 'onPress');

    expect(mockOnPress).not.toBeCalled();
  });
});
