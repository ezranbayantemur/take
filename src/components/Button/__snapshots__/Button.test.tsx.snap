// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button unit tests should render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#DB040E",
      "borderRadius": 10,
      "margin": 8,
      "opacity": 1,
      "padding": 16,
    }
  }
  testID="basetest_button_touchable"
>
  <Text
    style={
      {
        "color": "#FBFBFB",
        "fontFamily": "Lato-Bold",
        "fontSize": 15,
      }
    }
    testID="basetest_button_text"
  >
    test title
  </Text>
</View>
`;
